### ISSUE TITLE
The dim is not checked strictly for a model in GPU

### ISSUE THREAD

**Issue by WonderSeven** [pytorch/pytorch#68934]
## 🐛 Bug

Hi, I define a linear layer(input_dim=512, output_dim=10), when I create a tensor with shape [16, 522] and input it to the model, an inconsistent behavior happens in CPU and GPU environments.

In CPU only:
![image](https://user-images.githubusercontent.com/24557554/143549120-a68ed02a-7284-4f2d-892f-ed0b50398464.png)
It looks right, however, when I run this code in GPU, it passed

In GPU:
![image](https://user-images.githubusercontent.com/24557554/143549249-68c1c18b-957c-47f8-bb2d-91024307b736.png)

It outputs successfully. Wow, seems pytorch cut the dim to match them well? So if this is designed intentionally or other? Thx very much if someone can answer me. 

## To Reproduce

Here is my code:

```
import os
import torch
import torch.nn as nn


class Classifier(nn.Module):
    def __init__(self, input_dim, output_dim):
        super(Classifier, self).__init__()
        self.linear = nn.Linear(input_dim, output_dim)

    def forward(self, x):
        out = self.linear(x.view(x.size(0), -1))
        return out


if __name__ == '__main__':
    os.environ['CUDA_VISIBLE_DEVICES'] = '2'
    use_cuda = torch.cuda.is_available() # you can change here with False

    cla_func = Classifier(input_dim=512, output_dim=10)
    print(cla_func)
    if use_cuda:
        cla_func.cuda()

    with torch.no_grad():
        x = torch.rand([16, 512]).float()
        w = torch.rand([16, 10]).float()
        xw = torch.cat([x, w], dim=1)

        if use_cuda:
            xw = xw.cuda()

        print(xw.shape)
        y = cla_func(xw)
        print(y.shape)
```

Environments:
![image](https://user-images.githubusercontent.com/24557554/143553843-3839ff5b-89dd-4521-aa06-0603db23bdfb.png)

**Comment by ngimel** [comment]
Fixed in 1.9.1 and 1.10
